feat(aya-ebpf): Implement memmove

wdullaer opened the pull request:
The compiler will emit this function for certain operations, but aya currently does not provide an implementation.
This leads to ebpf loading failures as the kernel can't find the symbol when loading the program.

The implementation is based on https://github.com/rust-lang/compiler-builtins/blob/master/src/mem/mod.rs#L29-L40 and https://github.com/rust-lang/compiler-builtins/blob/master/src/mem/impls.rs#L128-L135 Only the simplest case has been implemented, none of the word optimizations, since memcpy also doesn't seem to have them.

vadorovsky commented:
I'm actually quite surprised that rust compiler-builtins don't optimize this implementation somehow (by loading bigger chunks and reducing iterations). I think in eBPF we could load with `u64` chunks. But anyway, that's not related this PR. :smiley:

vadorovsky reviewed (changes requested):
Thanks a lot for the change! :pray: 

Could you write a small integration tests which uses memmove? Scrolling back in Discord, I see that your motivation behind this is being able to do

```rust
unsafe { (*ip_hdr).dst_addr.in6_u.u6_addr8 = new_ip };
```

As it was resulting in a verifier error. I think it would be nice to have a test for that.

wdullaer commented:
It wasn't immediately clear to me where to add tests for this, which is why they're absent.

I'll take a stab at writing an integration test.
It might be hard to guarantee that we're using `memmove` though, since people reported on discord that using `clone` at the right place allowed them to avoid it.

vadorovsky commented:
I think if you are able to write a small program which triggers the error on your kernel, it's almost certainly going to be a reliable reproduces. Just avoid `clone()` in your example and make sure it fails with the current Aya codebase (without your changes). :)

Happy to help with the integration test.

The first step would be adding an `.rs` file with a regular aya-ebpf based program here:

https://github.com/aya-rs/aya/tree/main/test/integration-ebpf/src

Then including it here as a `[[ bin ]]` in `Cargo.toml`:

https://github.com/aya-rs/aya/blob/main/test/integration-ebpf/Cargo.toml

And writing a test on the user-space side. I think in your case just loading the program would be totally sufficient, no need to do any other actions.

https://github.com/aya-rs/aya/blob/main/test/integration-test/src/tests/load.rs#L285

I would literally just try something like:

```rust
#[test]
fn memmove() {
    let mut bpf = Ebpf::load(crate::TEST).unwrap();
    let prog: &mut SchedClassifier = bpf
        .program_mut("test_memmove")
        .unwrap()
        .try_into()
        .unwrap();

    prog.load().unwrap();
    assert_loaded("test_memmove");
```

I assumed `SchedClassifier` as a program type, since the case with overriding the packet data seems the easiest and TC has less limitations than XDP, at least for testing.

wdullaer commented:
I have added an integration test, and simplified the implementation of `memmove`.
The integration test will fail when the `memmove` code is commented out, it passes when `memmove` is present.
The verifier error produced by the integration tests is slightly different from what I saw previously, but I think it may be due to the lack of BTF info here.

I have copy/pasted some struct definitions from `network_types` to produce the test.
If you prefer I can also take a dependency on the library.

vadorovsky commented:
> I have added an integration test, and simplified the implementation of `memmove`. The integration test will fail when the `memmove` code is commented out, it passes when `memmove` is present.

Awesome, that's exactly what I had in mind. :slightly_smiling_face: 

> The verifier error produced by the integration tests is slightly different from what I saw previously, but I think it may be due to the lack of BTF info here.

What's exactly the error?

> I have copy/pasted some struct definitions from `network_types` to produce the test. If you prefer I can also take a dependency on the library.

I think I would be fine with adding a dependency on the library, it doesn't have any dependencies by default.

vadorovsky reviewed (changes requested):
The `lint` job is failing, because of public API changes.

@wdullaer Can you run `cargo xtask public-api --bless` and commit the result (preferably as a separate commit)? Also, before doing so, make sure you rebase.

I also have a slight preference towards importing the network-types crate instead of copying the types. I don't think it's going to make the build significantly slower. @alessandrod do you agree?

Otherwise, it looks very good, thanks for working on this! :pray:

vadorovsky reviewed (approved):
Amazing stuff, looks good!

I'm fine with keeping `*mut u8` (for consistency with the compiler-builtins impl and so).

tamird commented:
You'll need to rebase to fix the integration test failures.

vadorovsky commented:
@wdullaer Lint errors look legit. The integration test errors should be gone after a rebase.

wdullaer commented:
I'm sorry about the lint failures.
For some reason I can't get the lsp to work when in the integration test module. My dev/test cycle is not very efficient for this project at the moment.
Anyway, it should all be good now.
The integration tests pass on my VM now, apart from 2 tests which also fail on main.
